onmt/transforms/features.py

```diff
@@ -0,0 +1,94 @@
+from onmt.utils.logging import logger
+from onmt.transforms import register_transform
+from .transform import Transform, ObservableStats
+from onmt.constants import DefaultTokens, SubwordMarker
+from onmt.utils.alignment import subword_map_by_joiner, subword_map_by_spacer
+import re
+from collections import defaultdict
+
+
+
+@register_transform(name='filterfeats')
+class FilterFeatsTransform(Transform):
+    """Filter out examples with a mismatch between source and features."""
+
+    def __init__(self, opts):
+        super().__init__(opts)
+
+    @classmethod
+    def add_options(cls, parser):
+        pass
+
+    def _parse_opts(self):
+        pass
+
+    def apply(self, example, is_train=False, stats=None, **kwargs):
+        """Return None if mismatch"""
+
+        if 'src_feats' not in example:
+            # Do nothing
+            return example
+
+        for feat_name, feat_values in example['src_feats'].items():
+            if len(example['src']) != len(feat_values):
+                logger.warning(f"Skipping example due to mismatch between source and feature {feat_name}")
+                return None
+        return example
+
+    def _repr_args(self):
+        return ''
+
+
+@register_transform(name='inferfeats')
+class InferFeatsTransform(Transform):
+    """Infer features for subword tokenization."""
+
+    def __init__(self, opts):
+        super().__init__(opts)
+
+    @classmethod
+    def add_options(cls, parser):
+        """Avalilable options related to this Transform."""
+        group = parser.add_argument_group("Transform/InferFeats")
+        group.add("--reversible_tokenization", "-reversible_tokenization", default="joiner",
+                  choices=["joiner", "spacer"], help="Type of reversible tokenization applied on the tokenizer.")
+
+    def _parse_opts(self):
+        super()._parse_opts()
+        self.reversible_tokenization = self.opts.reversible_tokenization
+
+    def apply(self, example, is_train=False, stats=None, **kwargs):
+
+        if "src_feats" not in example:
+            # Do nothing
+            return example
+
+        # TODO: support joiner_new or spacer_new options. Consistency not ensured currently
+
+        if self.reversible_tokenization == "joiner":
+            word_to_subword_mapping = subword_map_by_joiner(example["src"])
+        else: #Spacer
+            # TODO: case markup
```

Comment:
?

Reply:
Currently `spacer` tokenization does not handle case markup placeholders.

Reply:
Not sure to understand. Are you talking about your code or the OpenNMT Tokenizer / `pyonmttok`? Because last time I checked spacer + case_markup works fine.
Couldn't we just use `subword_map_by_spacer` here?

Reply:
I've modified the `subword_map_by_joiner` function to accept case_markup tokens. What I didn't change was the `subword_map_by_spacer` function. It is pretty straightforward but I needs to be done. That's why I added the TODO. It will be done it the next commit.